ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -523,24 +528,104 @@ KEXT_STATIC int HandleFileOpOperation(
+        
+        // TODO(Mac): We need to handle failures to lookup the vnode.  If we fail to lookup the vnode
+        // it's possible that we'll miss notifications
+        errno_t toErr = vnode_lookup(newPath, 0 /* flags */, &currentVnode, context);
+        if (0 != toErr)
         {
-            PerfSample hardLinkSample(&perfTracer, PrjFSPerfCounter_FileOp_HardLinkCreated);
+            KextLog_Error("HandleFileOpOperation: vnode_lookup failed, errno %d for path '%s'", toErr, newPath);
+            goto CleanupAndReturn;
+        }
         
-            int kauthResult;
-            int kauthError;
-            if (!ProviderMessaging_TrySendRequestAndWaitForResponse(
-                    root,
-                    MessageType_KtoU_NotifyFileHardLinkCreated,
-                    currentVnode,
-                    vnodeFsidInode,
-                    newPath,
-                    pid,
-                    procname,
-                    &kauthResult,
-                    &kauthError))
+        // Don't expect named stream here as they can't be directly hardlinked or renamed, only the main fork can
+        assert(!vnode_isnamedstream(currentVnode));
+        
+        putCurrentVnode = true;
+        
+        bool isDirectory = (0 != vnode_isdir(currentVnode));
+        
+        VirtualizationRootHandle targetRoot, fromRoot;
+        pid_t pid;
+        bool messageTargetProvider =
+            ShouldHandleFileOpEvent(
+                &perfTracer,
+                context,
+                currentVnode,
+                nullptr, // don't pass path for target provider
+                action,
+                isDirectory,
+                &targetRoot,
+                &pid);
+        bool messageFromProvider =
+            ShouldHandleFileOpEvent(
+                &perfTracer,
+                context,
+                currentVnode,
+                fromPath,
+                action,
+                isDirectory,
+                &fromRoot,
+                &pid);
+        
+        if (!messageTargetProvider && !messageFromProvider)
+        {
+            goto CleanupAndReturn;
+        }
+        
+        FsidInode vnodeFsidInode = Vnode_GetFsidAndInode(currentVnode, context, true /* the inode is used for getting the path in the provider, so use linkid */);
+
+        char procname[MAXCOMLEN + 1];
+        proc_name(pid, procname, MAXCOMLEN + 1);
+
+        {
+            PerfSample sample(&perfTracer, PrjFSPerfCounter_FileOp_HardLinkCreated);
+
+            if (messageTargetProvider)
             {
-                goto CleanupAndReturn;
+                int kauthResult;
+                int kauthError = 0;
+                if (!ProviderMessaging_TrySendRequestAndWaitForResponse(
+                        targetRoot,
+                        MessageType_KtoU_NotifyFileHardLinkCreated,
+                        currentVnode,
+                        vnodeFsidInode,
+                        newPath,
+                        fromPath,
```

Comment:
I *think* we only want to send fromPath here if the providers match.  Otherwise we're sending a message for another Virtualization Root.

Reply:
> Currently the provider will fail to obtain a root-relative path for it, so the relativePath variable will be a nullptr and behave as expected

Yes, but we will print an error https://github.com/microsoft/VFSForGit/blob/57db0f04fa75be0ebb5b74a0d08ccada7c1bc9de/ProjFS.Mac/PrjFSLib/PrjFSLib.cpp#L1428

If we just null it out here no error

Reply:
@pmj, this is the last comment i'm waiting on before approving.  Let me know what you think!

Reply:
I've just pushed a new version that doesn't send any out-of-root paths with hardlink messages, whether that be "from" or "to" paths/vnodes - we'll see if it passes the functional tests or if that change will require some further work.